lib/writer/block-read-stream.js

```diff
@@ -0,0 +1,212 @@
+const fs = require('fs')
+const debug = require('debug')('block-read-stream')
+
+const CHUNK_SIZE = 64 * 1024
+const MIN_CHUNK_SIZE = 512
+
+/**
+ * @summary BlockReadStream
+ * @class
+ */
+class BlockReadStream extends stream.Readable {
+  /**
+   * @summary BlockReadStream constructor
+   * @param {Object} [options] - options
+   * @param {Number} [options.fd] - file descriptor
+   * @param {String} [options.path] - file path
+   * @param {String} [options.flags] - file open flags
+   * @param {Number} [options.mode] - file mode
+   * @param {Number} [options.start] - start offset in bytes
+   * @param {Number} [options.end] - end offset in bytes
+   * @param {Boolean} [options.autoClose] - automatically close the stream on end
+   * @example
+   * new BlockReadStream()
+   */
+  constructor (options) {
+    options = Object.assign({}, BlockReadStream.defaults, options)
+    options.objectMode = true
+
+    debug('block-read-stream %j', options)
+
+    super(options)
+
+    this.fs = options.fs
+    this.fd = options.fd
+    this.path = options.path
+    this.flags = options.flags
+    this.mode = options.mode
+    this.end = options.end || Infinity
+    this.autoClose = options.autoClose
+
+    this.position = options.start || 0
+    this.bytesRead = 0
+
+    this.closed = false
+    this.destroyed = false
+
+    this.once('end', function () {
+      if (this.autoClose) {
+        this.close()
+      }
+    })
+
+    /**
+     * @summary onRead handler
+     * @param {Error} error - error
+     * @param {Number} bytesRead - bytes read
+     * @param {Buffer} buffer - resulting buffer
+     * @example
+     * fs.read(fd, buffer, 0, length, position, onRead)
+     */
+    this._onRead = (error, bytesRead, buffer) => {
+      if (!error && bytesRead !== buffer.length) {
+        error = new Error(`Bytes read mismatch: ${bytesRead} != ${buffer.length}`)
+      }
+
+      if (error) {
+        if (this.autoClose) {
+          this.destroy()
+        }
+        this.emit('error', error)
+        return
+      }
+
+      this.bytesRead += bytesRead
+      this.push(buffer)
+    }
+
+    this.open()
+  }
+
+  /**
+   * @summary Read a chunk from the source
+   * @private
+   * @example
+   * // not to be called directly
+   */
+  _read () {
+    // Wait for file handle to be open
+    if (this.fd == null) {
+      this.once('open', () => {
+        this._read()
+      })
+      return
+    }
+
+    const toRead = this.end - this.position
+
+    if (toRead <= 0) {
+      this.push(null)
+      return
+    }
+
+    const length = Math.min(CHUNK_SIZE, Math.max(MIN_CHUNK_SIZE, toRead))
+    const buffer = Buffer.alloc(length)
+
+    this.fs.read(this.fd, buffer, 0, length, this.position, this._onRead)
+    this.position += length
+  }
+
+  /**
+   * @summary Open a handle to the file
+   * @private
+   * @example
+   * this.open()
+   */
+  open () {
+    debug('open')
+
+    if (this.fd != null) {
+      this.emit('open', this.fd)
```

Comment:
> Should there be code to prevent .open being called twice? 

No, it's made to be called even with an open fd, so that the "open" event will be emitted correctly etc.

> ...and is/should it be permissible to call .open on a stream after you've .closed it? (and if not should that be checked for?)

This is handled by Node core streams (which we're inheriting from) – anything after close/end/destroy will cause "read/write after close" errors